silkworm/rpc/core/account_dumper.cpp

```diff
@@ -55,90 +54,97 @@ Task<DumpAccounts> AccountDumper::dump_accounts(
+        dump_account.code_hash = account->code_hash;
+        dump_account.root = kZeroHash;
 
-        if (account->incarnation > 0 && account->code_hash == kEmptyHash) {
-            const auto storage_key{db::storage_prefix(address.bytes, account->incarnation)};
-            auto code_hash{co_await transaction_.get_one(db::table::kPlainCodeHashName, storage_key)};
-            if (code_hash.length() == kHashLength) {
-                std::memcpy(dump_account.code_hash.bytes, code_hash.data(), kHashLength);
+        if (account->code_hash != kZeroHash && !exclude_code) {
+            db::kv::api::GetLatestQuery query_code{
+                .table = db::table::kCodeDomain,
+                .key = db::account_domain_key(address)};
+
+            const auto code = co_await transaction_.get_latest(std::move(query_code));
+            if (!code.value.empty()) {
+                dump_account.code = code.value;
             }
         }
-        if (!exclude_code) {
-            auto code = co_await state_reader.read_code(address, dump_account.code_hash);
-            dump_account.code.swap(code);
-        }
         dump_accounts.accounts.insert(std::pair<evmc::address, DumpAccount>(address, dump_account));
     }
 
-    co_return;
+    if (!exclude_storage) {
+        co_await load_storage(block_num, dump_accounts);
+    }
+
+    co_return dump_accounts;
 }
 
 Task<void> AccountDumper::load_storage(BlockNum block_num, DumpAccounts& dump_accounts) {
-    SILK_TRACE << "block_num " << block_num << " START";
+    SILK_TRACE << "block_number " << block_num << " START";
     StorageWalker storage_walker{transaction_};
-    evmc::bytes32 start_location{};
+    const auto txn_number = co_await transaction_.first_txn_num_in_block(block_num);
+
     for (auto& it : dump_accounts.accounts) {
         auto& address = it.first;
         auto& account = it.second;
 
+        auto to = db::code_domain_key(address);
+        increment(to);
+
+        db::kv::api::DomainRangeQuery query{
+            .table = db::table::kStorageDomain,
+            .from_key = db::code_domain_key(address),
+            .to_key = to,
+            .timestamp = txn_number,
+            .ascending_order = true};
+
+        auto paginated_result = co_await transaction_.range_as_of(std::move(query));
+        auto sit = co_await paginated_result.begin();
         std::map<Bytes, Bytes> collected_entries;  // TODO(canepat) switch to ByteView?
-        StorageWalker::AccountCollector collector = [&](const evmc::address& /*address*/, ByteView loc, ByteView data) {
+
+        while (const auto value = co_await sit.next()) {
+            if (value->second.empty())
+                continue;
+
             if (!account.storage.has_value()) {
                 account.storage = Storage{};
             }
             auto& storage = *account.storage;
-            storage[to_bytes32(loc)] = data;
+            auto loc = value->first.substr(20);
```

Comment:
- it's safer checking `value.first` size before extracting a substring
- use `kAddressLength` constant instead of hard-coded value
- `auto loc` -> `const auto loc`

Reply:
ok